### ISSUE TITLE
[Impeller] Proposal: Support rendering analytically clean "squircles" and "rectellipses".

### ISSUE THREAD

**Issue by chinmaygarde** [flutter/flutter#139321]
Branching out this from https://github.com/flutter/flutter/issues/13914#issuecomment-1832599538 which discusses such paths on iOS (which may or may not be approximations). This issue tracks being able to render analytically clean superellipses with Impeller.

How such an implementation may interact with the existing path APIs will need to be tackled too.

**Comment by sethladd** [comment]
I really like supporting Squircles. It's the "one more thing" we can add, and it shows the attention to detail that we want to be known for.

**Comment by bernaferrari** [comment]
The issue so far, if I had to guess, is that everybody wants to make it "perfect" and it is (probably) impossible to achieve perfectness, so someone needs to figure out the right amount of trade-offs between magic numbers, understanding advanced linear algebra or reverse engineering things, and there are many possibilities, but so far no one has said "mine is better, let's merge it" since they all have trade-offs (even Figma has).

**Comment by rydmike** [comment]
The reason why I made my Squircle study was to compare existing solutions and discover their shortcomings.

This comment mentions some of them  https://github.com/flutter/flutter/issues/13914#issuecomment-1832640718 but for full insights reading the study is recommended https://github.com/rydmike/squircle_study

**Comment by sethladd** [comment]
Thank you for cross-linking!

**Comment by Piinks** [comment]
Is https://github.com/flutter/flutter/issues/91523 also a dupe of this issue?

**Comment by bernaferrari** [comment]
Yes

**Comment by chinmaygarde** [comment]
This issue came up in recent conversations and we discussed it during the Impeller weekly yesterday.

Impeller separates geometry and color source concerns (flutter.dev/go/impeller-geometry) and it should be possible to generate a mesh directly. The thinking is that this will allow for the creation of superellipses without having to resort to Bézier approximation. Our understanding is that this is hard and misses edge cases.

An implementation has been prototyped in https://github.com/flutter/engine/pull/54562 by Jonah.

Unlike implementations that use Bézier approximation, components of this shape cannot be expressed as paths. Instead, users will typically have to draw the shape using a direct canvas call.

The engine team also doesn’t have a good intuition for how to parameterize these shapes in a manner that is developer friendly. You’ll notice the parameters in the demo are just from the superellipse equation. But these may not be intuitive to developers.

In the demo, a degree of 4 should give you a squircle. We could also expose more arguments (different exponents perhaps) to make even more novel shapes. But we don’t know what use that would be.

The engine team would like some guidance on:
* Do folks care about clean superellipses or is bézier approximation good enough? I don’t have the expertise to approximate a superellipse with bézier curves but drawing one directly is possible with Impeller.
  * If bézier approximation is good enough, there is nothing more to do in the engine and packages/framework can do it just fine.
* Are the curve components not being able to be added as path components (ui.Path) a deal breaker?
* Are there ways to parameterize the shape in a manner that is more developer friendly?

https://github.com/user-attachments/assets/ffd1ff28-d79c-42e2-a61a-e947d2bc5d11

**Comment by flar** [comment]
Also, does stroking a superellipse/rectellipse/squircle matter?

**Comment by bernaferrari** [comment]
> Do folks care about clean superellipses?

No

> is bézier approximation good enough

Yes (figma is doing that, it has some "glitches" on large or specific values, but no one ever complained. It also has glitches on StarBorder, I think more glitches than Flutter)

> Are the curve components not being able to be added as path components (ui.Path) a deal breaker?

I don't know, but if it works in ShapeDecoration should be fine.



> Also, does stroking a superellipse/rectellipse/squircle matter?

Yes.

**Comment by dkwingsmt** [comment]
Progress: 
- [x] Basic asset geometry (https://github.com/flutter/engine/pull/56726)
- [ ] Asymmetric corner radius
- [ ] Stroke
- [ ] Clips
- [ ] Integration with DL/Canvas

I'll keep this list updated.

**Comment by jonahwilliams** [comment]
@dkwingsmt time to update the list? I believe this can be closed out now right?

**Comment by dkwingsmt** [comment]
You're right! I believe this can be closed! Thank you @jonahwilliams for the immense support throughout the project, as well as much help from @flar and @gaaclarke !
